File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys 
from collections import deque
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
dx = [0, 0, -1, 1]
dy = [1, -1, 0, 0]

h, w= map(int, input().split())
graph = [list(map(str, input())) for _ in range(h)]
def bfs():
    visited = [[[0]*2 for _ in range(w)] for _ in range(h)]
    queue = deque()
    queue.append((0,0,0))
    visited[0][0][0] = 1

    while queue:
        x, y, z = queue.popleft()
        if x==h-1 and y==w-1:
            return visited[x][y][z]
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0<=nx<h and 0<=ny<w and visited[nx][ny][z]==0:
                if graph[nx][ny]=='0':
                    visited[nx][ny][z]=visited[x][y][z]+1
                    queue.append((nx,ny,z))
                elif graph[nx][ny]=='1' and z==0:
                    visited[nx][ny][z+1]=visited[x][y][z]+1
                    queue.append((nx,ny,z+1))

    return -1
print(bfs())

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Queue struct {
	x []int
	y []int
	z []int
}

func (q *Queue) Enqueue(x, y, z int) {
	q.x = append(q.x, x)
	q.y = append(q.y, y)
	q.z = append(q.z, z)
}

func (q *Queue) Dequeue() (x, y, z int) {
	if len(q.x) <= 0 {
		return -1, -1, -1
	} else {
		x = q.x[0]
		y = q.y[0]
		z = q.z[0]
		if len(q.x) == 1 {
			q.x = []int{}
			q.y = []int{}
			q.z = []int{}
		} else {
			q.x = q.x[1:]
			q.y = q.y[1:]
			q.z = q.z[1:]
		}
	}
	return
}

func (q *Queue) Len() int {
	return len(q.x)
}

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var w, h int
	var cx, cy, cz int
	var nx, ny int
	var res int
	var s string
	res = -1
	fmt.Fscanf(reader, "%d %d\n", &h, &w)
	dx := []int{0, 0, -1, 1}
	dy := []int{1, -1, 0, 0}
	var graph = make([][]string, h)
	var visited = make([][][]int, h)
	for i := 0; i < h; i++ {
		visited[i] = make([][]int, w)
		for j := 0; j < w; j++ {
			visited[i][j] = make([]int, 2)
		}
		fmt.Fscanf(reader, "%s\n", &s)
		graph[i] = strings.Split(s, "")
	}
	var q = new(Queue)
	visited[0][0][0] = 1
	q.Enqueue(0, 0, 0)
	for q.Len() > 0 {
		cx, cy, cz = q.Dequeue()
		if cx == h-1 && cy == w-1 {
			res = visited[cx][cy][cz]
			break
		}
		for i := 0; i < 4; i++ {
			nx = cx + dx[i]
			ny = cy + dy[i]
			if 0 <= nx && nx < h && 0 <= ny && ny < w && visited[nx][ny][cz] == 0 {
				if graph[nx][ny] == "0" {
					visited[nx][ny][cz] = visited[cx][cy][cz] + 1
					q.Enqueue(nx, ny, cz)
				} else if graph[nx][ny] == "1" && cz == 0 {
					visited[nx][ny][cz+1] = visited[cx][cy][cz] + 1
					q.Enqueue(nx, ny, cz+1)
				}
			}
		}
	}
	fmt.Fprintf(writer, "%d", res)
}
